lib/image_optim/cache.rb

```diff
@@ -0,0 +1,68 @@
+require 'digest/sha1'
+require 'fspath'
+
+class ImageOptim
+  # Handles image cache
+  class Cache
+    def initialize(image_optim, workers_by_format)
+      return unless image_optim.cache_dir
+      @cache_dir = FSPath.new(image_optim.cache_dir)
+      @cache_workers = image_optim.cache_workers
+      @options_by_format = Hash[workers_by_format.map do |k, v|
+        [k, v.map(&:inspect).sort.join(', ')]
+      end]
+      @bins_by_format = Hash[workers_by_format.map do |k, v|
+        [k, v.map(&:used_bins).flatten!.map! do |sym|
+          bin = image_optim.resolve_bin!(sym)
+          "#{bin.name}[#{bin.digest}]"
+        end.sort!.uniq.join(', ')]
+      end]
+    end
+
+    def read(original)
+      return unless @cache_dir
+
+      cached = @cache_dir / digest(original)
+      return unless File.exist?(cached)
+
+      cached
+    end
+
+    def write(original, optimized)
+      return unless @cache_dir
+
+      digest = digest(original)
+      cached = @cache_dir / digest
+      FileUtils.mkdir_p(File.dirname(cached))
+
+      if optimized
+        t = Time.now.strftime('%Y%m%d')
+        tmp = @cache_dir / "#{digest}#{t}-#{$PID}-#{rand(0x100000000).to_s(36)}"
```

Comment:
While this name will be most likely unique, I think it would still be better to use `Tempfile` through `FSPath` like:
```ruby
FSPath.temp_file_path(something, @cache_dir) do |tmp|
  # move and rename
end
```
`something` could have been just digest, but it already contains slash, which will create a problem for `Tempfile`